From 0fc67ff3ce0aaa7d48cf5e9c7154fe3bfbef74f8 Mon Sep 17 00:00:00 2001
From: "Vargas, Hernan" <hernan.vargas@intel.com>
Date: Thu, 13 Mar 2025 04:08:03 +0000
Subject: [PATCH] SCSY-210415 Fix multiple acc config issue

Creating multiple cluster configs to configure multiple accelerators
causes the controller to detect false configuration changes.
Fix deep equal comparison by sorting accelerator objects before
comparing.

Signed-off-by: Hernan Vargas <hernan.vargas@intel.com>
---
 .../sriovfecclusterconfig_controller.go       | 17 ++++++++++++++---
 .../sriovvrbclusterconfig_controller.go       | 18 +++++++++++++++---
 pkg/daemon/daemon_fec.go                      | 19 +++++++------------
 3 files changed, 36 insertions(+), 18 deletions(-)

diff --git a/controllers/sriovfec/sriovfecclusterconfig_controller.go b/controllers/sriovfec/sriovfecclusterconfig_controller.go
index 2cfbd8e..a28b139 100644
--- a/controllers/sriovfec/sriovfecclusterconfig_controller.go
+++ b/controllers/sriovfec/sriovfecclusterconfig_controller.go
@@ -139,7 +139,7 @@ func (r *SriovFecClusterConfigReconciler) synchronizeNodeConfigSpec(ncc NodeConf
 
 	newNodeConfig := copyWithEmptySpec(ncc.SriovFecNodeConfig)
 
-	// Use orederedmap for iteration
+	// Use orderedmap for iteration
 	for _, pciAddress := range acceleratorConfigContext.Keys() {
 		cc, _ := acceleratorConfigContext.Get(pciAddress)
 		pf := sriovfecv2.PhysicalFunctionConfigExt{
@@ -157,13 +157,24 @@ func (r *SriovFecClusterConfigReconciler) synchronizeNodeConfigSpec(ncc NodeConf
 		newNodeConfig.Spec.PhysicalFunctions = append(newNodeConfig.Spec.PhysicalFunctions, pf)
 	}
 
-	// copy latest known drainSkip from NodeConfig for cleanup
+	// Copy latest known drainSkip from NodeConfig for cleanup
 	if acceleratorConfigContext.Len() == 0 {
 		newNodeConfig.Spec.DrainSkip = ncc.Spec.DrainSkip
 	}
 
+	// Sort the physical functions by PCI address to ensure consistent order
+	sort.Slice(newNodeConfig.Spec.PhysicalFunctions, func(i, j int) bool {
+		return newNodeConfig.Spec.PhysicalFunctions[i].PCIAddress < newNodeConfig.Spec.PhysicalFunctions[j].PCIAddress
+	})
+	sort.Slice(currentNodeConfig.Spec.PhysicalFunctions, func(i, j int) bool {
+		return currentNodeConfig.Spec.PhysicalFunctions[i].PCIAddress < currentNodeConfig.Spec.PhysicalFunctions[j].PCIAddress
+	})
+
 	if !equality.Semantic.DeepEqual(newNodeConfig.Spec, currentNodeConfig.Spec) {
-		r.Log.Info("Node Config Changed")
+		r.Log.WithFields(logrus.Fields{
+			"CurrentSpec": currentNodeConfig.Spec,
+			"NewSpec":     newNodeConfig.Spec,
+		}).Info("Node Config Changed")
 		return r.Update(context.TODO(), newNodeConfig)
 	}
 	return nil
diff --git a/controllers/sriovvrb/sriovvrbclusterconfig_controller.go b/controllers/sriovvrb/sriovvrbclusterconfig_controller.go
index 35aaf08..3958416 100644
--- a/controllers/sriovvrb/sriovvrbclusterconfig_controller.go
+++ b/controllers/sriovvrb/sriovvrbclusterconfig_controller.go
@@ -127,7 +127,7 @@ func (r *SriovVrbClusterConfigReconciler) synchronizeNodeConfigSpec(ncc NodeConf
 
 	newNodeConfig := copyWithEmptySpec(ncc.SriovVrbNodeConfig)
 
-	// Use orederedmap for iteration
+	// Use orderedmap for iteration
 	for _, pciAddress := range acceleratorConfigContext.Keys() {
 		cc, _ := acceleratorConfigContext.Get(pciAddress)
 		pf := vrbv1.PhysicalFunctionConfigExt{
@@ -146,15 +146,27 @@ func (r *SriovVrbClusterConfigReconciler) synchronizeNodeConfigSpec(ncc NodeConf
 		newNodeConfig.Spec.PhysicalFunctions = append(newNodeConfig.Spec.PhysicalFunctions, pf)
 	}
 
-	// copy latest known drainSkip from NodeConfig for cleanup
+	// Copy latest known drainSkip from NodeConfig for cleanup
 	if acceleratorConfigContext.Len() == 0 {
 		newNodeConfig.Spec.DrainSkip = ncc.Spec.DrainSkip
 	}
 
+	// Sort the physical functions by PCI address to ensure consistent order
+	sort.Slice(newNodeConfig.Spec.PhysicalFunctions, func(i, j int) bool {
+		return newNodeConfig.Spec.PhysicalFunctions[i].PCIAddress < newNodeConfig.Spec.PhysicalFunctions[j].PCIAddress
+	})
+	sort.Slice(currentNodeConfig.Spec.PhysicalFunctions, func(i, j int) bool {
+		return currentNodeConfig.Spec.PhysicalFunctions[i].PCIAddress < currentNodeConfig.Spec.PhysicalFunctions[j].PCIAddress
+	})
+
 	if !equality.Semantic.DeepEqual(newNodeConfig.Spec, currentNodeConfig.Spec) {
-		r.Log.Info("Node Config Changed")
+		r.Log.WithFields(logrus.Fields{
+			"CurrentSpec": currentNodeConfig.Spec,
+			"NewSpec":     newNodeConfig.Spec,
+		}).Info("Node Config Changed")
 		return r.Update(context.TODO(), newNodeConfig)
 	}
+
 	return nil
 }
 
diff --git a/pkg/daemon/daemon_fec.go b/pkg/daemon/daemon_fec.go
index 03a4714..6174447 100644
--- a/pkg/daemon/daemon_fec.go
+++ b/pkg/daemon/daemon_fec.go
@@ -83,21 +83,16 @@ func (r *FecNodeConfigReconciler) Reconcile(_ context.Context, req ctrl.Request)
 		return requeueLater()
 	}
 
-	if r.isCardUpdateRequired(sfnc, detectedInventory) {
-
-		if err := r.updateStatus(sfnc, metav1.ConditionFalse, ConfigurationInProgress, "Configuration started"); err != nil {
-			return requeueNowWithError(err)
-		}
+	if err := r.updateStatus(sfnc, metav1.ConditionFalse, ConfigurationInProgress, "Configuration started"); err != nil {
+		return requeueNowWithError(err)
+	}
 
-		if err := r.configureNode(sfnc); err != nil {
-			r.log.WithError(err).Error("error occurred during configuring node")
-			return requeueNowWithError(r.updateStatus(sfnc, metav1.ConditionFalse, ConfigurationFailed, err.Error()))
-		} else {
-			return requeueLaterOrNowIfError(r.updateStatus(sfnc, metav1.ConditionTrue, ConfigurationSucceeded, "Configured successfully"))
-		}
+	if err := r.configureNode(sfnc); err != nil {
+		r.log.WithError(err).Error("error occurred during configuring node")
+		return requeueNowWithError(r.updateStatus(sfnc, metav1.ConditionFalse, ConfigurationFailed, err.Error()))
 	}
 
-	return requeueLater()
+	return requeueLaterOrNowIfError(r.updateStatus(sfnc, metav1.ConditionTrue, ConfigurationSucceeded, "Configured successfully"))
 }
 
 /*****************************************************************************
-- 
2.48.1.windows.1

